### DISCUSSION TITLE
How to avoid creating path with over and lensPath?

### DISCUSSION THREAD

**Discussion by robations** [ramda/ramda#3362] [Q&A]
Can anyone point me in the right direction here?

```
const fn = x => x === undefined ? x : "bar";
over(lensPath(["foo"]), fn, {});
// actual output: { foo: undefined }
// desired output: {}
```

In other words, if the path doesn't exist, I don't want any changes to the object.

I'm guessing this could be done by handling `fn -> undefined` as a special case, or by detecting the path does not exist and never calling fn.

Would/could I create a customised `over` or `lensPath`?

**Comment by robations** [comment]
...similar to the newish modifyPath behaviour (which I've only just noticed exists), but interested to see if there is a lens-like way to do this.

**Comment by jlissner** [comment]
Maybe something like this is what you're looking for

```javascript
const fn = x => x === undefined ? x : "bar";
const notNil = complement(isNil); // just an example predicate
const overWhen = curry((pred, lp, fn, obj) =>
  when(
    pipe(view(lp), pred), // passing the value from view to the predicate
    over(lp, fn) // does the "over" when the value passes the predicate
  )(obj)
)
const overWhenExists = overWhen(notNil) // for your specific request

overWhenExists(lensPath(["foo"]), fn, {}); // {}
```

**Reply by robations** [reply]
Thanks Joe. This would work for my use case.

An niggle (somewhat academic) is that you still can’t differentiate between "path doesn't exist" and undefined values. If you change the example `fn`:

```
const fn2 = x => x === undefined ? "poo" : "bar";
const overWhenExists = overWhen(notNil) // for your specific request

overWhenExists(lensPath(["foo"]), fn, {});
// actual: { foo: "poo" }
// desired: {}
```

I wonder if this is actually not possible because lensPath abstracts the view. It could maybe be solved by returning some kind of symbol to differentiate "path doesn't exist" but this would require a modified lensPath to handle the symbol.
